elixir/lib/app_identity/plug.ex

```diff
@@ -285,14 +407,70 @@ if Code.ensure_loaded?(Plug.Conn) do
       end
     end
 
+    defp get_name(options) do
+      case Keyword.get(options, :name, :app_identity) do
+        value when is_atom(value) -> value
+        _ -> raise AppIdentityError, :plug_name_invalid
+      end
+    end
+
     defp get_headers(options) do
-      headers = Keyword.get(options, :headers)
+      case Keyword.get(options, :headers) do
+        nil ->
+          nil
 
-      if !is_list(headers) || Enum.empty?(headers) do
-        raise AppIdentityError, :plug_headers_required
+        [] ->
+          raise AppIdentityError, :plug_header_invalid
+
+        headers when not is_list(headers) ->
+          raise AppIdentityError, :plug_header_invalid
+
+        headers ->
+          if duplicate_headers?(headers) do
+            raise AppIdentityError, :plug_header_invalid
+          end
+
+          Enum.map(headers, &parse_option_header/1)
+      end
+    end
+
+    defp get_header_groups(options) do
+      case Keyword.get(options, :header_groups) do
+        nil ->
+          nil
+
+        groups when not is_map(groups) ->
+          raise AppIdentityError, :plug_header_groups_invalid
+
+        groups ->
+          if Enum.empty?(groups) do
+            raise AppIdentityError, :plug_header_groups_invalid
+          end
+
+          invalid_names? =
+            groups
+            |> Map.keys()
+            |> Enum.any?(fn v -> !is_binary(v) end)
+
+          empty_groups? = Enum.any?(groups, &match?({_, []}, &1))
+
+          if invalid_names? || empty_groups? || duplicate_headers?(Map.values(groups)) do
+            raise AppIdentityError, :plug_header_groups_invalid
+          end
+
+          Map.new(groups, fn {name, headers} ->
+            {name, Enum.map(headers, &parse_option_header/1)}
+          end)
       end
+    end
 
-      Enum.map(headers, &parse_option_header/1)
+    defp duplicate_headers?(headers) do
```

Comment:
I [benchmarked](https://gist.github.com/halostatue/4407dde79bb9c584abcf083f6b83b1ea) an even *simpler* implementation (using `--`), but it’s 2x slower than either the `group_by/filter` approach or the `frequencies` approach. The use of `frequencies/2` is slower in all tests than `group_by/filter`, but minimally so (~5%) except in the pathological case when it's faster. In all cases `frequencies` uses 40% less memory than `group_filter`, even better than `--` in some cases.

Depending on the way that AppIdentity.Plug gets used, that minimal difference may make a difference in response times. When used with a Plug.Builder (e.g., `plug AppIdentity.Plug, …`), the duplicate detection and such is compiled in. If used *inline* (e.g., `AppIdentity.Plug.call(conn, …)`), the duplicate detection and such is evaluated at runtime (like in our tests, except for the one with the explicit module compilation).

Regardless, I’m accepting your suggested implementation because it gets rid of `|| Kernel.not()`.